### DISCUSSION TITLE
[NodeJs] How to choose correct value of Max Buffer Size in Skywalking nodejs Agent?

### DISCUSSION THREAD

**Discussion by br-arruda** [apache/skywalking#11145] [Q&A] [answered]
Hi,

I'm planning to use [this agent](https://github.com/apache/skywalking-nodejs), I read the documentation in project and at certain point I read that:

> endpoints which are not recorded due to exceeding SW_AGENT_MAX_BUFFER_SIZE all propagate their ignored status downstream to any other endpoints they may call. If that endpoint is running the Node Skywalking agent then regardless of its ignore settings it will not be recorded since its upstream parent was not recorded.

I have some large requests ongoing in my system, I need to monitor those too but now I'm afraid they won't be caught by the agent. 

So I want to adjust this buffer according to my needs. The doc only states that the default value is `1000` but it doesn't explain its unit or what consumes this limit. Is it in bytes? And what part of incoming request is being buffered? All of it? Just the headers? 

**What measurements** should I analyze about my requests in order to get into an optimal size for this buffer?

Thanks in advance!

**Comment by wu-sheng** [comment]
About buffer size, there is no easy answer. It depends on how quickly the gRPC client could forward the data to the remote OAP server. The only recommendation is based on your perf testing.

**Reply by br-arruda** [reply]
Thanks for you response! 

Okay, can I assume that the field `SW_AGENT_MAX_BUFFER_SIZE` I mentioned is measured in bytes?

About the perf testing you said, can you give me an example? I already have configured [Self-Observability](https://skywalking.apache.org/docs/main/next/en/setup/backend/dashboards-so11y) for Skywalking, but I don't know yet what indicators should I pay more attention to. Probably you are talking about some metric or testing scenario that I'm not aware of, and I would be grateful if you could briefly explain to me what is this perf testing or how I could measure the performance of SW or it's agents. I found no doc about it yet 😞

**Reply by wu-sheng** [reply, marked as answer]
> Okay, can I assume that the field SW_AGENT_MAX_BUFFER_SIZE I mentioned is measured in bytes?

Not that complex. I believe all this kind of buffer sizes in SkyWalking means the size of the number of span segments in the reporting queue/buffer. I think you can see this metric on the self-obs dashboard.

<img width="929" alt="image" src="https://github.com/apache/skywalking/assets/5441976/7eb1eef7-6d28-49ad-956d-f44c72ed1d96">

Notice, this self-obs is per node of OAP,  you need to verify the load is balanced or not.

**Reply by br-arruda** [reply]
Cool. Thanks! I will keep testing my deployment here, and seek for the optimal load balance distribution among the OAP pods.

I will close this question and keep testing my deployment here, seeking for the optimal setup for Skywalking stack. 

I suggest you to include in the parameters readme section of  [NodeJS Agent](https://github.com/apache/skywalking-nodejs#set-up-nodejs-agent) that `SW_AGENT_MAX_BUFFER_SIZE` relate to the amount of span segments, helping other new users to understand what it does.

**Reply by wu-sheng** [reply]
Feel free to open a PR to update.
